5-WebApp-AuthZ/5-2-Groups/Services/MicrosoftGraph-Rest/GraphHelper.cs

```diff
@@ -0,0 +1,168 @@
+﻿using Microsoft.AspNetCore.Authentication.OpenIdConnect;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Graph;
+using System;
+using System.Collections.Generic;
+using System.IdentityModel.Tokens.Jwt;
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using WebApp_OpenIDConnect_DotNet.Infrastructure;
+
+namespace WebApp_OpenIDConnect_DotNet.Services
+{
+    public class GraphHelper
+    {
+        /// <summary>
+        /// This method inspects the claims collection created from the ID or Access token issued to a user and returns the groups that are present in the token . If it detects groups overage,
+        /// the method then makes calls to Microsoft Graph to fetch the group membership of the authenticated user.
+        /// </summary>
+        /// <param name="context">TokenValidatedContext</param>
+        public static async Task<List<string>> GetSignedInUsersGroups(TokenValidatedContext context)
```

Comment:
>GetSignedInUsersGroups [](start = 47, length = 22)

1. check if token overgae occured, if yes
 1.1 check for the existance of the session var ( if (_httpContextAccessor.HttpContext.Session.Keys.Contains("groupClaims")))
  1.1.1 if session var the groups of the user fetched earlier, return the groups
  1.1.1 if session var does not have the groups of the user, fetch the groups from graph and populate the session var
2. return the groups array for the caller